testinfra/utils/ansible_runner.py

```diff
@@ -18,150 +18,183 @@
+        self.host_list = host_list
 
     @classmethod
     def get_runner(cls, inventory):
-        try:
-            return cls._runners[inventory]
-        except KeyError:
+        # stores a copy of the runner in a dict keyed by inv
+        if inventory not in cls._runners:
             cls._runners[inventory] = cls(inventory)
-            return cls._runners[inventory]
+        return cls._runners[inventory]
+
+    def fetch_inventory(self, host=None):
+        '''Helper function for ansible-inventory'''
+
+        cmd = 'ansible-inventory -i %s' % self.host_list
+        if host is not None:
+            cmd += ' --host=%s' % host
+        else:
+            cmd += ' --list'
+
+        env = os.environ.copy()
+        env['ANSIBLE_NOCOLOR'] = "1"
+        p = subprocess.Popen(
+            cmd,
+            shell=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=env
+        )
+        (so, se) = p.communicate()
 
+        if p.returncode != 0:
+            msg = 'ansible-inventory failed: %s' % se
+            raise AnsibleInventoryException(msg)
 
-class Callback(ansible.plugins.callback.CallbackBase):
+        inv = json.loads(so)
 
-    def __init__(self, *args, **kwargs):
-        self.result = {}
-        super(Callback, self).__init__(*args, **kwargs)
+        return inv
 
-    def runner_on_ok(self, host, result):
-        self.result = result
+    def get_hosts(self, pattern=None):
+        '''Return a list of host names from inventory via the pattern'''
+        inv = self.fetch_inventory()
+        return list(inv['_meta']['hostvars'].keys())
 
-    def runner_on_failed(self, host, result, ignore_errors=False):
-        self.result = result
+    def get_variables(self, host, refresh=True):
+        '''Get a mixture of inventory vars and magic vars'''
 
-    # pylint: disable=no-self-use
-    def runner_on_unreachable(self, host, result):
-        raise RuntimeError(
-            'Host {} is unreachable: {}'.format(
-                host, pprint.pformat(result)),
+        if host not in self.variables or refresh:
+
+            _vars = self.variables.get(host, {})
+
+            # inventory vars
+            _vars.update(self.fetch_inventory(host=host))
+
+            # this is a hack to get the magic vars
+            res = self.run(host, 'debug', 'var=hostvars')
+            _vars.update(res.get('hostvars', {}).get(host, {}))
+
+            # one of the unit tests insist this should be returned
+            _vars['inventory_hostname'] = host
+            self.variables[host] = _vars
+
+        return self.variables[host]
+
+    def run(self, host, module_name, module_args=None, **kwargs):
+        '''Invokes a single module on a single host and returns dict results'''
+
+        # runner must have a directory based payload
+        data_dir = tempfile.mkdtemp(prefix='runner.data.%s.' \
+            % datetime.datetime.now().isoformat().replace(':', '-'))
```

Comment:
I'd suggest to use ``tempfile.TemporaryDirectory()`` context manager instead to avoid leaving the directory in case of Exception.

Reply:
Also why using a private data dir for ansible runner ? I think users expect to have all their files (envvars, groupvars, ansible.cfg, inventory, etc) availables in testinfra tests.

Reply:
This is another architecture carryover from how runner is an evolution of the internal code for tower/awx and then isolated nodes. The idea being (if i'm doing it justice), the tower job template data gets exposed to a proot'ed/bublewrap'ed environment where all the "isolation" for the execution of ansible occurs.

Reply:
I'll also point out that a lot of the inputs can be provided as direct inputs to the run method (obviating the need for a private data dir).

Runner does need (and will get) a mechanism for not needing a private data dir or artifacts dir.